### PR TITLE
Start of MailChimp v3 changes

### PR DESCRIPTION
**Pull request by charleswaddell**
## Summary
On April 2nd, 2019 MailChimp is shutting down a bunch of the older versions of the API which Deliverance is currently using. These changes swap in the Mailchimp API 3.0 along with removing a bunch of old code that is no longer being used. Instead of converting a bunch of old code I decided it would be faster to remove the old code instead of converting it.

It turns out that after I removed the dead code the only big things that we needed to convert were the `subscribe`, `unsubscribe`, `batchSubscribe`, `batchUnsubscribe`, `getMemberInfo`, and `isAvailable` methods in `DeliveranceMailChimpList`. These are the methods use on the newsletter sign up and unsubscribe pages.

The main part of these changes are in `DeliveranceMailChimpList::callClientMethod` and `DeliveranceMailChimpList::handleClientErrors`. These have been updated to use the new MailChimp client and handle the new errors that come along with it.

You can see the new API being used in  `DeliveranceMailChimpList::getMemberInfo()`. RESTful URLs are now used instead of the method calls from before. There appear to be many less error types to deal with so I've divided them into three exception types - client errors (400s), server errors (500s) and timeouts.

The client and servers errors can be tested by [passing along a special header](https://developer.mailchimp.com/documentation/mailchimp/guides/get-started-with-mailchimp-api-3/#test-error-handling) with your request. For timeout testing I added the following line to the `MailChimp::makeRequest()` method.

```
curl_setopt($ch, CURLOPT_PORT, 81);
```

## Testing

Pick any of the RAP sites and test that the newsletter sign up and unsubscribe pages work. Also the mailing list updater should also be tested. Currently - we are running into the issue described [here](https://github.com/drewm/mailchimp-api/issues/267) when doing batch processing but with the change applied the new code works.

## Switching a site to the new code

- [ ] Replace `"silverorange/mailchimp_api": "^2.0.0",` with `"drewm/mailchimp-api": "^2.5"` in the `composer.json`.
- [ ] Remove references to the `DeliveranceCampaignFactory` from the site singleton.
- [ ] Remove the newsletter admin component.
- [ ] Remove the segment cache updater from the list of cron scripts.
- [ ] Remove DB tables for `MailingListCampaignSegment`, `MailingListMemberCache`, `MailingListMemberUpdaterCache`, and `MailChimpOrderQueue`
- [ ] [Convert](https://github.com/silverorange/deliverance/pull/51#user-content-shortnames) mailing list shortnames

## <a name="shortnames">Converting mailing list interests shortnames</a>

The `MailingListInterest` table in each site's DB will need to be updated for the new version of the API. In order to tell MailChimp what interests a user should belong to we need the interests ID which isn't currently in the DB table.

The old version of the API used the `shortname` but this has been replaced by an ID. For my testing I've used the [Mailchimp API Playground](https://us1.api.mailchimp.com/playground/) to look up the new values for the `shortname` and `group_shortname` fields. Once those are in place updating interests will work just like before. Any hardcoded strings in code that reference the old shortname will need to be updated as well.

## Helpful Links
 - [Mailchimp API 3.0 Error Glossary](https://developer.mailchimp.com/documentation/mailchimp/guides/error-glossary/)
 - [Mailchimp API 3.0 Reference](https://developer.mailchimp.com/documentation/mailchimp/reference/overview/)
- [Mailchimp API 1.3 Reference](https://web.archive.org/web/20150918233634/https://apidocs.mailchimp.com/api/1.3/listsubscribe.func.php) from the wayback machine. Very useful for seeing what the old API calls provided.

### CODE REVIEW

**Comment by gauthierm** [comment]
Is there still value in the queued list updates model?

File: Deliverance/DeliveranceMailChimpList.php [hunk 1]
```diff
@@ -1,8 +1,11 @@
 <?php
 
+use \DrewM\MailChimp\MailChimp;
```
Comment (gauthierm) [inline Deliverance/DeliveranceMailChimpList.php]: PHP namespace imports should not have a leading `\`

File: Deliverance/DeliveranceMailChimpList.php [hunk 2]
```diff
@@ -1081,749 +559,118 @@ public function getInterests()
...
-	public function addOrder(array $order_info)
-	{
-		$success = false;
-
-		try {
-			$parameters = array(
-				$order_info,
+			// Server exceptions
+			if ($last_response['headers']['http_code'] >= 500) {
+				$error = json_decode($last_response['body']);
```
Comment (gauthierm) [inline Deliverance/DeliveranceMailChimpList.php]: Add error handling for failure to decode response JSON. It's possible if MailChimp goes down that we'd get a 503 backend fetch failure that is not JSON-encoded.

**Comment by charleswaddell** [comment]
> Is there still value in the queued list updates model?

What do you mean? Does it still make sense for use to have a queue of subscribes and unsubscribes?

**Comment by m-mitchell** [comment]
- I tested this out on Essentials and I can add myself to the mailing list, but the mailing list interests don't come through. (I also tested on Kendra's testbed to make sure it wasn't just broken on the site level, and that worked.) 
- What's the correct replacement for the `update` methods? I assumed `subscribe` because I saw a note on `batchSubscribe` that said existing addresses would be updated.

**Comment by charleswaddell** [comment]
> I tested this out on Essentials and I can add myself to the mailing list, but the mailing list interests don't come through. (I also tested on Kendra's testbed to make sure it wasn't just broken on the site level, and that worked.)

Change b538cea has fixed this issue. Each site will need their mailing list interests in the DB updated as well. See [this section]() in the PR description for an explanation of what needs to be done.

> What's the correct replacement for the update methods? I assumed subscribe because I saw a note on batchSubscribe that said existing addresses would be updated.

The update method has been added back in b538cea. The list updater is also reading from the queue again and dealing with those request properly.

File: Deliverance/DeliveranceMailChimpList.php [hunk 3]
```diff
@@ -1081,749 +698,127 @@ public function getInterests()
...
-		DeliveranceMailChimpCampaign $campaign
-	) {
-		$segment_options = $campaign->getSegmentOptions();
-
-		if ($segment_options != null) {
-			if ($this->getSegmentSize($segment_options) == 0) {
-				throw new DeliveranceCampaignException(
-					'Campaign Segment Options return no members');
+			$error = json_decode($last_response['body']);
+			if (!is_object($error)) {
```
Comment (gauthierm) [inline Deliverance/DeliveranceMailChimpList.php]: use `$error !=== null`

**Comment by charleswaddell** [comment]
[ch4545]

File: Deliverance/DeliveranceList.php [hunk 4]
```diff
@@ -396,73 +297,48 @@ public function queueSubscribe($address, array $info, $send_welcome = false)
...
-		return self::QUEUED;
-	}
-
-	// }}}
-	// {{{ public function queueUnsubscribe()
+	// {{{ public function queueUpdate()
 
 	/**
-	 * Enqueues an unsubscribe request for this list
+	 * Enqueues a update subscription request for this list
```
Comment (m-mitchell) [inline Deliverance/DeliveranceList.php]:
```suggestion
	 * Enqueues an update subscription request for this list
```

File: Deliverance/DeliveranceList.php [hunk 5]
```diff
@@ -396,73 +297,48 @@ public function queueSubscribe($address, array $info, $send_welcome = false)
...
+		$info = serialize($info);
+
 		$transaction = new SwatDBTransaction($this->app->db);
 		try {
 
 			$sql = sprintf(
-				'select count(1) from MailingListUnsubscribeQueue
-				where email = %s and instance %s %s',
+				'select count(1) from MailingListUpdateQueue
+				where email = %s and info = %s and instance %s %s',
```
Comment (m-mitchell) [inline Deliverance/DeliveranceList.php]: 👍

File: Deliverance/DeliveranceListUpdater.php [hunk 6]
```diff
@@ -344,11 +249,11 @@ private function getQueuedSubscribes($with_welcome)
 	// }}}
 	// {{{ protected function getQueuedUpdates()
 
-	protected function getQueuedUpdates()
+	private function getQueuedUpdates()
```
Comment (m-mitchell) [inline Deliverance/DeliveranceListUpdater.php]: Why private vs protected?

File: Deliverance/DeliveranceMailChimpList.php [hunk 7]
```diff
@@ -348,66 +168,46 @@ protected function initListMergeArrayMap()
...
-					$result = DeliveranceList::INVALID;
-					break;
-
-				default:
-					throw $e;
-				}
+			} catch (DeliveranceMailChimpServerException $e) {
+				$queue_request = true;
+			} catch (DeliveranceMailChimpClientException $e) {
+				$result = DeliveranceList::INVALID;
```
Comment (m-mitchell) [inline Deliverance/DeliveranceMailChimpList.php]: While testing, I got this error for other reasons than having an invalid email. Not sure if this is something we can check for or if it's worth checking for right now. I got the error for trying to submit invalid interests.

**Pull request merged**